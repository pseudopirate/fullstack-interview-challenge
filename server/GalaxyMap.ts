import { Planet } from '../models/common';

interface Route {
    from: Planet['code'];
    to: Planet['code'][];
}

interface ChildNode {
    parent?: ChildNode;
    name: Planet['code'];
}

class GalaxyMap {
    private routes: Record<string, Route> = {};

    addRoute(from: string, to: string) {
        if (!this.routes[from]) { // add origin to the map of existing nodes
            this.routes[from] = {from, to: []};
        }

        if (!this.routes[from].to.includes(to)) {
            this.routes[from].to.push(to);
        }

        if (!this.routes[to]) { // add dest to the map of existing nodes
            this.routes[to] = {from: to, to: []};
        }
    }

    getRoute(from: string) {
        return this.routes[from];
    }

    findShortestPath(from: string, to: string) { // breadth first search

        // path doesn't exist if nodes doesn't exist
        if (!this.routes[from] || !this.routes[to]) { 
            return [];
        }

        const visited: ChildNode[] = [];

        const nodes: ChildNode[] = [{name: from}];

        while (nodes.length > 0) {
            const node = nodes.shift() as ChildNode;
            
            const children = this.routes[node.name].to;
            
            children.forEach((child) => {
                const isInVisited = visited.find(({name}) => name === child);
                const isInNodes = nodes.find(({name}) => name === child);

                if (!isInVisited && !isInNodes) {
                    nodes.push({
                        name: child,
                        parent: node,
                    });
                }
            });

            visited.push(node);


            if (node.name === to) {
                break;
            }
        }

        // follow the links from destination to root to find total path
        let currentNode = visited.pop();

        if (currentNode?.name !== to) { // path not found
            return [];
        }
        
        const path: string[] = [];

        while (currentNode) {
            path.unshift(currentNode.name);

            currentNode = currentNode.parent;
        }

        return path;
    }

    isDeadEnd(node: string) {
        const route = this.routes[node];

        return route && route.to.length > 0 ? false : true;
    }
}

export default GalaxyMap;
