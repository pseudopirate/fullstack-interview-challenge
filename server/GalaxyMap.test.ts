import GalaxyMap from './GalaxyMap';

describe('GalaxyMap', () => {
    describe('addRoute(from: string, to: string', () => {
        it('should add routes', () => {
            const gm = new GalaxyMap();

            gm.addRoute('TEST1', 'TEST2');
            gm.addRoute('TEST1', 'TEST3');
            gm.addRoute('TEST2', 'TEST3');

            expect(gm).toMatchSnapshot();
        });

        it('should add routes without duplicates', () => {
            const gm = new GalaxyMap();

            gm.addRoute('TEST1', 'TEST2');
            gm.addRoute('TEST1', 'TEST2');

            expect(gm.getRoute('TEST1')).toEqual({
                from: 'TEST1',
                to: ['TEST2'],
            });
        });
    });

    describe('findShortestPath(from: string, to: string)', () => {
        describe('Graph 1', () => {
            /*
                 a
                /|\
               b c d
              /| | | 
             a f e i 
               |
               e
            */
            const gm = new GalaxyMap();

            gm.addRoute('a', 'b');
            gm.addRoute('a', 'c');
            gm.addRoute('a', 'd');
            gm.addRoute('b', 'f');
            gm.addRoute('b', 'a');
            gm.addRoute('c', 'e');
            gm.addRoute('d', 'i');
            gm.addRoute('f', 'e');
    
            it('should match snapshot', () => {
                expect(gm).toMatchSnapshot(); // check if all nodes are saved in routes
            });
            
            it('should return path', () => {
                expect(gm.findShortestPath('a', 'e')).toEqual(['a', 'c', 'e']);
            });
        });

        describe('Graph 2', () => {
            /*
                 a     e
                /|\    | 
               b c d   f 
            */
            const gm = new GalaxyMap();

            gm.addRoute('a', 'b');
            gm.addRoute('a', 'c');
            gm.addRoute('a', 'd');
            gm.addRoute('e', 'f');

            it('should return empty path', () => {
                expect(gm.findShortestPath('a', 'f')).toEqual([]);
            });

            it('should return empty path', () => {
                expect(gm.findShortestPath('s', 'c')).toEqual([]);
            });

            it('should return empty path', () => {
                expect(gm.findShortestPath('a', 'v')).toEqual([]);
            });
        });
    });

    describe('isDeadEnd(node: string)', () => {
        const gm = new GalaxyMap();

        gm.addRoute('a', 'b');

        it('should be truhty', () => {
            expect(gm.isDeadEnd('b')).toBeTruthy();
            expect(gm.isDeadEnd('c')).toBeTruthy();
        });

        it ('should be falsy', () => {
            expect(gm.isDeadEnd('a')).toBeFalsy();
        });
    });
});
