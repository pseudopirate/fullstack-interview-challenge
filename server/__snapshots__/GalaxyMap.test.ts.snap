// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalaxyMap addRoute(from: string, to: string should add routes 1`] = `
GalaxyMap {
  "routes": Object {
    "TEST1": Object {
      "from": "TEST1",
      "to": Array [
        "TEST2",
        "TEST3",
      ],
    },
    "TEST2": Object {
      "from": "TEST2",
      "to": Array [
        "TEST3",
      ],
    },
    "TEST3": Object {
      "from": "TEST3",
      "to": Array [],
    },
  },
}
`;

exports[`GalaxyMap findShortestPath(from: string, to: string) Graph 1 should match snapshot 1`] = `
GalaxyMap {
  "routes": Object {
    "a": Object {
      "from": "a",
      "to": Array [
        "b",
        "c",
        "d",
      ],
    },
    "b": Object {
      "from": "b",
      "to": Array [
        "f",
        "a",
      ],
    },
    "c": Object {
      "from": "c",
      "to": Array [
        "e",
      ],
    },
    "d": Object {
      "from": "d",
      "to": Array [
        "i",
      ],
    },
    "e": Object {
      "from": "e",
      "to": Array [],
    },
    "f": Object {
      "from": "f",
      "to": Array [
        "e",
      ],
    },
    "i": Object {
      "from": "i",
      "to": Array [],
    },
  },
}
`;
